src/crypter.rs

```diff
@@ -0,0 +1,145 @@
+use aes_gcm::{
+    aead::{AeadCore, AeadInPlace, KeyInit, OsRng},
+    Aes128Gcm as AesGcm, Error, KeySizeUser, Nonce,
+};
+use base64::{engine::general_purpose::STANDARD, DecodeError, Engine};
+use std::{
+    collections::VecDeque,
+    fmt::{self, Debug, Formatter},
+    iter,
+    str::FromStr,
+    sync::Arc,
+};
+use typenum::marker_traits::Unsigned;
+
+pub type Key = aes_gcm::Key<AesGcm>;
+
+#[derive(Clone)]
+pub struct Crypter(Arc<CrypterInner>);
+
+#[derive(thiserror::Error, Debug, Clone)]
+pub enum CrypterParseError {
+    #[error(transparent)]
+    Base64(#[from] DecodeError),
+
+    #[error("incorrect key length, must be {} bytes after base64 decode", <AesGcm as KeySizeUser>::KeySize::USIZE)]
+    IncorrectLength,
+
+    #[error("at least one key needed")]
+    Missing,
+}
+
+impl FromStr for Crypter {
+    type Err = CrypterParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut keys = s
+            .split(',')
+            .map(|s| {
+                STANDARD
```

Comment:
Any reason not to use `NO_PAD`? (base64 padding is only necessary/useful if multiple b64 values need to be concatenated together, then parsed apart again; I don't think we need to do that.)

Less important, but any reason not to use the URL-safe alphabet?

Reply:
I have no opinion on this. I feel like this comes up every time we base64 something. I'm happy to switch it to URL_SAFE_NO_PAD or whatever is easiest for the generating side. Humans should never see these values so it doesn't matter how they're encoded

Reply:
Padding is not sufficient to allow concatenation of multiple base64-encoded values together. As a counterexample, note that `AAAA` is a valid encoding of three null bytes in all of base64, base64url, base64 without padding, and base64url without padding. Thus, it is not possible to unambiguously decode multiple concatenated encoded values from an input that consists of `A` repeated a multiple of four times.

[RFC 4648 section 3.2](https://datatracker.ietf.org/doc/html/rfc4648#section-3.2) is very brief, but it seems to suggest that this is instead useful when the protocol employing base64 may be unclear about the "end of file".

Weighing in the other direction, base64 with padding is both recommended by RFC 4648 and better supported by tooling.

Reply:
addressed in 2a40071fb31f6d98bef09fe22b3fecc403f36025